browser/ui/browser_commands.cc

```diff
@@ -178,19 +178,20 @@ base::FilePath GetDefaultFilepathForBookmarkExport() {
  * This class is responsible for showing a file dialog to the user for selecting
  * the location to save exported bookmarks.
  *
- * @note The lifetime of this class is tied to the FileSelected dialog. It will
- * be automatically deleted when the dialog is closed, a file is selected, or
- * the dialog is cancelled.
+ * @note The lifetime of this class is managed through weak pointers to ensure
+ * safe cleanup when the async bookmark writing operation completes.
  */
 class BookmarksExportListener : public ui::SelectFileDialog::Listener {
  public:
   explicit BookmarksExportListener(Profile* profile)
       : profile_(profile),
-        file_selector_(ui::SelectFileDialog::Create(this, nullptr)) {}
+        file_selector_(ui::SelectFileDialog::Create(this, nullptr)),
+        weak_factory_(this) {}
   void FileSelected(const ui::SelectedFileInfo& file, int index) override {
-    bookmark_html_writer::WriteBookmarks(profile_, file.file_path,
-                                         base::DoNothing());
-    delete this;
+    bookmark_html_writer::WriteBookmarks(
+        profile_, file.file_path,
+        base::BindOnce(&BookmarksExportListener::OnBookmarksWritten,
+                       weak_factory_.GetWeakPtr()));
```

Comment:
actually, let's fix it the other way.

1. keep the original impl, but add `file_selector_->ListenerDestroyed()` before `delete this;`. This way the dangling raw_ptr in SelectFileDialog will be unset which will prevent the dangling check.
2. add `FileSelectionCanceled() override` and do the same there:
```
file_selector_->ListenerDestroyed();
delete this;
```